clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java

```diff
@@ -37,16 +38,20 @@
     private final List<KafkaMetric> metrics;
     private final MetricConfig config;
     private final Time time;
+    private volatile long lastRecordTime;
+    private final long expireInactiveSensorTimeMillis;
 
-    Sensor(Metrics registry, String name, Sensor[] parents, MetricConfig config, Time time) {
+    Sensor(Metrics registry, String name, Sensor[] parents, MetricConfig config, Time time, long expireInactiveSensorTimeSeconds) {
         super();
         this.registry = registry;
         this.name = Utils.notNull(name);
         this.parents = parents == null ? new Sensor[0] : parents;
-        this.metrics = new ArrayList<KafkaMetric>();
-        this.stats = new ArrayList<Stat>();
+        this.metrics = new ArrayList<>();
+        this.stats = new ArrayList<>();
         this.config = config;
         this.time = time;
+        this.expireInactiveSensorTimeMillis = TimeUnit.MILLISECONDS.convert(expireInactiveSensorTimeSeconds, TimeUnit.SECONDS);
+        this.lastRecordTime = time.milliseconds();
```

Comment:
I thought about that a bit. Setting it to the current time lets the sensor get garbage collected if no record() is ever called.